hs_labels/models.py

```diff
@@ -0,0 +1,876 @@
+# - u.ulabels.unlabel_resource(r, label)
+#   removes a label; there can be many labels.
+# - u.ulabels.clear_resource_labels(r)
+#   removes all labels for a resource
+# - u.ulabels.favorite_resource(r)
+#   favorites a resource
+# - u.ulabels.unfavorite_resource(r)
+#   removes a favorite
+# - u.ulabels.file_resource(r, folder_name)
+#   adds a folder name for a resource
+# - u.ulabels.unfile_resource(r)
+#   remove folder name
+#
+# and the reporting functions
+# - u.ulabels.labeled_resources
+#   A queryset of resources that are labeled.
+# - u.ulabels.filed_resources (trivial, not yet implemented, available tomorrow)
+#   A queryset of resources that are filed in folders
+# - u.ulabels.favorite_resources (trivial, not yet implemented, available tomorrow)
+#   A queryset of resources that have been favorited
+# - u.ulabels.get_resources_with_label(label)
+#   Get a queryset of resources possessing a specific label.
+# - u.ulabels.get_resources_in_folder(folder)
+#   Get a queryset of resources in a specific folder.
+
+# For a BaseResource r, this also adds a subobject rlabels that reports on labels for resources
+# -	r.rlabels.get_folder(u)
+# - r.rlabels.get_labels(u)
+# - r.rlabels.is_favorite(u) (trivial, not yet implemented, tomorrow)
+
+from django.contrib.auth.models import User
+from django.db import models
+#from django.db.models import Q
+# from django.conf import settings
+#from pprint import pprint
+import re
+
+from hs_core.models import BaseResource
+
+# from hs_core.models import BaseResource
+# for debugging, create a stub BaseResource class
+# fields will be induced into this class via Many-Many relations
+# class BaseResource(models.Model):
+#     title = models.TextField(default='nothing')
+#     pass
+
+######################################
+######################################
+# Labeling and annotation subsystem
+######################################
+######################################
+
+# A "label" is something that is assigned by a person other than the original creator.
+# It is not part of metadata. Labels include
+# 1) making something a favorite.
+# 2) assigning a folder
+# 3) assigning a keyword
+
+######################################
+# Exceptions specific to access control:
+# b) Usage Exception: bad parameters
+######################################
+class HSLException(Exception):
+    """
+    Generic Base Exception class for HSLLabel class.
+
+    *This exception is a generic base class and is never directly raised.*
+    See subclass HSLUsageException for details.
+    """
+    pass
+
+class HSLUsageException(HSLException):
+    """
+    Exception class for parameter usage problems.
+
+    This exception is raised when a routine is called with invalid parameters.
+    This includes references to non-existent resources, groups, and users.
+
+    This is typically a programming error and should be entered into
+    issue tracker and resolved.
+
+    *Catching this exception is not recommended.*
```

Comment:
OK, but programming errors still shouldn't break a page if we can help it. If there is an error, we should be able to let the user know something went wrong, and get that information to the right people to fix. Raising it to a 500 response isn't helping the user or us.